tsl/src/compression/recompress.c

```diff
@@ -0,0 +1,745 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <parser/parse_coerce.h>
+#include <parser/parse_relation.h>
+#include <utils/inval.h>
+#include <utils/snapmgr.h>
+#include <utils/syscache.h>
+#include <utils/typcache.h>
+
+#include "compression.h"
+#include "create.h"
+#include "guc.h"
+#include "hypercore/hypercore_handler.h"
+#include "hypercore/utils.h"
+#include "recompress.h"
+#include "ts_catalog/array_utils.h"
+#include "ts_catalog/chunk_column_stats.h"
+#include "ts_catalog/compression_settings.h"
+
+static bool fetch_uncompressed_chunk_into_tuplesort(Tuplesortstate *tuplesortstate,
+													Relation uncompressed_chunk_rel);
+static void update_current_segment(CompressedSegmentInfo *current_segment, TupleTableSlot *slot,
+								   int nsegmentby_cols);
+static void create_segmentby_scankeys(CompressionSettings *settings, Relation index_rel,
+									  Relation compressed_chunk_rel, ScanKeyData *index_scankeys);
+static void create_orderby_scankeys(CompressionSettings *settings, Relation index_rel,
+									Relation compressed_chunk_rel, ScanKeyData *orderby_scankeys);
+static void update_segmentby_scankeys(TupleTableSlot *uncompressed_slot,
+									  CompressedSegmentInfo *current_segment, int num_segmentby,
+									  ScanKey index_scankeys);
+static void update_orderby_scankeys(TupleTableSlot *uncompressed_slot,
+									CompressedSegmentInfo *current_segment, int num_segmentby,
+									int num_orderby, ScanKey orderby_scankeys);
+static enum Batch_match_result match_tuple_batch(TupleTableSlot *compressed_slot, int num_orderby,
+												 ScanKey orderby_scankeys, bool *nulls_first);
+static bool check_changed_group(CompressedSegmentInfo *current_segment, TupleTableSlot *slot,
+								int nsegmentby_cols);
+static void recompress_segment(Tuplesortstate *tuplesortstate, Relation compressed_chunk_rel,
+							   RowCompressor *row_compressor);
+/*
+ * Recompress an existing chunk by decompressing the batches
+ * that are affected by the addition of newer data. The existing
+ * compressed chunk will not be recreated but modified in place.
+ *
+ * 0 uncompressed_chunk_id REGCLASS
+ * 1 if_not_compressed BOOL = false
+ */
+Datum
+tsl_recompress_chunk_segmentwise(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_not_compressed = PG_ARGISNULL(1) ? true : PG_GETARG_BOOL(1);
```

Comment:
The default is true here while in the function comment it is stated as false

Reply:
Will address this in a follow-up PR since its current state and has very little to do with the change here.